src/xmpp/xmpp-im/xmpp_mamtask.cpp

```diff
@@ -0,0 +1,178 @@
+#include "xmpp_mamtask.h"
+
+using namespace XMLHelper;
+using namespace XMPP;
+
+class MAMTask::Private {
+public:
+    Jid       j;
+    QDateTime from;
+    QDateTime to;
+    bool      allowMUCArchives;
+    int mamPageSize; // TODO: this is the max page size for MAM request. Should be made into a config option
+    int       mamMaxMessages; // maximum mam pages total, also should be config
+    bool flipPages;
+    bool backwards;
+
+    QList<QDomElement> archive;
+    int                messagesFetched;
+    bool               metadataFetched;
+    QString firstID;
+    QString lastID;
+    QString lastArchiveID;
+};
+
+MAMTask::MAMTask(Task *parent) : Task(parent) { d = new Private; }
+MAMTask::~MAMTask() { delete d; }
+
+const QList<QDomElement> &MAMTask::archive() const { return d->archive; }
+
+void MAMTask::get(const Jid &j, const QDateTime &from = QDateTime(), const QDateTime &to = QDateTime(),
+                  const bool allowMUCArchives = true, int mamPageSize = 10, int mamMaxMessages = 100, bool flipPages = true, bool backwards = true)
+{
+    d->archive         = QList<QDomElement>();
+    d->messagesFetched = 0;
+    d->metadataFetched = false;
+
+    d->j                = j;
+    d->from             = from;
+    d->to               = to;
+    d->allowMUCArchives = allowMUCArchives;
+    d->mamPageSize = mamPageSize;
+    d->mamMaxMessages   = mamMaxMessages;
+    d->flipPages = flipPages;
+    d->backwards = backwards;
+}
+
+XData MAMTask::makeMAMFilter() {
+    XData::FieldList fl;
+
+    XData::Field type;
+    type.setType(XData::Field::Field_Hidden);
+    type.setVar("FORM_TYPE");
+    type.setValue(QStringList("urn:xmpp:mam:2"));
+    fl.append(type);
+
+    XData::Field with;
+    type.setType(XData::Field::Field_JidSingle);
+    type.setVar("with");
+    type.setValue(QStringList(d->j.bare()));
+    fl.append(with);
+
+    XData::Field includeGroupchat;
+    includeGroupchat.setType(XData::Field::Field_Boolean);
+    includeGroupchat.setVar("include-groupchat");
+    includeGroupchat.setValue(QStringList(d->allowMUCArchives ? "true" : "false"));
+    fl.append(includeGroupchat);
+
+    if (d->from.isValid()) {
+        XData::Field start;
+        start.setType(XData::Field::Field_TextSingle);
+        type.setVar("start");
+        d->from.setTimeSpec(Qt::UTC);
+        type.setValue(QStringList(d->from.toString()));
+        fl.append(start);
+    }
+
+    if (d->to.isValid()) {
+        XData::Field end;
+        end.setType(XData::Field::Field_TextSingle);
+        type.setVar("end");
+        d->to.setTimeSpec(Qt::UTC);
+        type.setValue(QStringList(d->to.toString()));
+        fl.append(end);
+    }
+
+    // TODO: ID filters?
```

Comment:
most likely those will be required. 
imagine a situation you chatted with using Conversations a lot and everything was written to MAM.
Then you switched to Psi (which was offline till that moment) and loaded just the most recent history from server.
Then your buddy sends you a message with a quote which has message ID referring your message. Psi doesn't have this message in local database and it doesn't know when it was sent, so it needs to fetch it from MAM by ID and likely some messages before and after it to show in the log.